src/Features/Core/Portable/PullMemberUp/MembersPuller.cs

```diff
@@ -0,0 +1,408 @@
+            // It will just pull abstract declaration up to destination.
+            // But if the member is abstract itself, it will still be removed.
+            foreach (var analysisResult in result.MemberAnalysisResults)
+            {
+                if (!analysisResult.MakeDeclarationAtDestinationAbstract)
+                {
+                    foreach (var syntax in syntaxMap[analysisResult.Member])
+                    {
+                        var originalMemberEditor = editorMap[syntax.SyntaxTree];
+                        originalMemberEditor.RemoveNode(originalMemberEditor.Generator.GetDeclaration(syntax));
+                    }
+                }
+            }
+
+            // Change the destination to abstract class if needed.
+            var destinationEditor = editorMap[destinationSyntaxNode.SyntaxTree];
+            if (!result.Destination.IsAbstract &&
+                result.MemberAnalysisResults.Any(analysis => analysis.Member.IsAbstract || analysis.MakeDeclarationAtDestinationAbstract))
+            {
+                var modifiers = DeclarationModifiers.From(result.Destination).WithIsAbstract(true);
+                newDestination = destinationEditor.Generator.WithModifiers(newDestination, modifiers);
+            }
+
+            destinationEditor.ReplaceNode(destinationSyntaxNode, (syntaxNode, generator) => newDestination);
+            return solutionEditor.GetChangedSolution();
+        }
+
+        private ISymbol GetAbstractMemberSymbol(ISymbol member)
+        {
+            if (member.IsAbstract)
+            {
+                return member;
+            }
+
+            var modifier = DeclarationModifiers.From(member).WithIsAbstract(true);
+            if (member is IMethodSymbol methodSymbol)
+            {
+                return CodeGenerationSymbolFactory.CreateMethodSymbol(methodSymbol, modifiers: modifier);
+            }
+            else if (member is IPropertySymbol propertySymbol)
+            {
+                return CodeGenerationSymbolFactory.CreatePropertySymbol(propertySymbol, modifiers: modifier);
+            }
+            else if (member is IEventSymbol eventSymbol)
+            {
+                return CodeGenerationSymbolFactory.CreateEventSymbol(eventSymbol, modifiers: modifier);
+            }
+            else
+            {
+                throw ExceptionUtilities.UnexpectedValue(member);
+            }
+        }
+
+        private async Task<(EditorMap editorMap, SyntaxMap syntaxMap)> InitializeEditorMapsAndSyntaxMapAsync(
+            PullMembersUpAnalysisResult result,
+            Solution solution,
+            SolutionEditor solutionEditor,
+            SyntaxNode destinationSyntaxNode,
+            CancellationToken cancellationToken)
+        {
+            // Members and destination may come from different documents,
+            // EditorMap is used to save and group all the editors will be used.
```

Comment:
The SolutionEditor already maintains a document -> DocumentEditor map, albeit not keyed by a syntax tree. Can we reuse?

Reply:
This is a great api for simplify code here.👍